docs/lang_spec/Cadence Language Specification.md

```diff
@@ -0,0 +1,5798 @@
+
+---
+
+### 1. Character Set
+
+Cadence source code is written using the **Unicode** character set (UTF-8 encoding), which allows for a wide range of characters from various languages. However, the core lexical elements, such as identifiers, operators, and punctuation, are primarily restricted to the ASCII subset of Unicode.
+
+- **Allowed Characters**: 
+  - Letters (a-z, A-Z)
+  - Digits (0-9)
+  - Special symbols used in operators and syntax, such as `+`, `-`, `*`, `/`, `=`, `{`, `}`, `(`, `)`, `[`, `]`, etc.
+  - Whitespace characters: space (`U+0020`), tab (`U+0009`), newline (`U+000A`), and carriage return (`U+000D`).
+
+- **Unicode Support**:
+  - While the core syntax is based on ASCII characters, Unicode is fully supported in **string literals** and **comments**.
+  - Identifiers may include Unicode characters in the ranges of letters and numbers but should generally follow best practices for naming and clarity.
+
+---
+
+### 2. Identifiers
+
+Identifiers in Cadence are used to name **variables**, **constants**, **functions**, **contracts**, **types**, and other user-defined entities. Identifiers must follow specific rules to ensure consistency and avoid conflicts with reserved keywords.
+
+- **Rules for Naming Identifiers**:
+  - Identifiers must start with a **letter** (either uppercase or lowercase) or an **underscore (`_`)**. 
+  - After the first character, identifiers may contain **letters**, **digits (0-9)**, or underscores (`_`).
+  - Example: `balance`, `transactionFee`, `_internalState`.
+
+- **Case Sensitivity**: 
+  - Identifiers in Cadence are **case-sensitive**. For example, `Token` and `token` are treated as distinct identifiers.
+
+- **Valid Characters**:
+  - Identifiers can include ASCII letters (`A-Z`, `a-z`), digits (`0-9`), and underscores (`_`).
+  - Identifiers cannot contain spaces or special characters like `!`, `@`, `#`, etc.
+
+- **Special Identifiers**:
+  - Identifiers starting with an underscore (`_`) are typically used for private or internal entities. Although allowed, their use should follow common programming conventions.
+
+---
+
+### 3. Keywords and Reserved Words
+
+Cadence has a set of **keywords** that are reserved for special syntactic purposes. These keywords cannot be used as identifiers in programs, as they have predefined meanings in the language.
+
+- **Keywords**:
+  - `access`, `as`, `attachment`, `auth`, `break`, `case`, `catch`, `const`, `continue`, `contract`, `create`, `default`, `destroy`, `else`, `emit`, `entitlement`, `enum`, `event`, `execute`, `export`, `false`, `final`, `finally`, `for`, `fun`, `goto`, `guard`, `if`, `import`, `in`, `include`, `init`, `interface`, `internal`, `is`, `let`, `mapping`, `move`, `native`, `nil`, `post`, `pre`, `prepare`, `priv`, `pub`, `repeat`, `require`, `requires`, `resource`, `result`, `return`, `self`, `static`, `struct`, `switch`, `throw`, `throws`, `transaction`, `true`, `try`, `typealias`, `var`, `where`, `while`.
```

Comment:
`pub` and `priv` got removed. Maybe update the list of keywords in the current parser. There's also the internal `info` command which can dump the hard keywords: `go run ./cmd/info dump-hard-keyword`

Reply:
Useful tool! `pub` and `priv` are still dumped. Should we removed them? Or are they useful for backwards compatibility?